src/dynamics/time_integration.jl

```diff
@@ -301,6 +278,347 @@ function first_timesteps!(
+    if N != 4
+        @warn "ABBA variant only for 4-cycle, using variant A"
+        return weight_coefficient(NCycleLorenzA(), k, N, step_counter)
+    end
+    
+    pattern_position = (step_counter ÷ N) % 4
+    if pattern_position == 0 || pattern_position == 3
+        return weight_coefficient(NCycleLorenzA(), k, N, step_counter)
+    else
+        return weight_coefficient(NCycleLorenzB(), k, N, step_counter)
+    end
+end
+
+# ============================================================================
+# LORENZ N-CYCLE HELPER KERNELS
+# ============================================================================
+
+"""$(TYPEDSIGNATURES)
+Accumulate weighted tendency: G = w*F_E + (1-w)*G (Hotta et al. 2016, Eq. 19)"""
+function accumulate_tendency!(
+    G::LowerTriangularArray,
+    F_explicit::LowerTriangularArray,
+    w::Real,
+    L::NCycleLorenz{NF},
+) where NF
+    @boundscheck size(G) == size(F_explicit) || throw(BoundsError())
+    
+    w_NF = convert(NF, w)
+    
+    launch!(
+        architecture(G), 
+        SpectralWorkOrder, 
+        size(G), 
+        accumulate_tendency_kernel!, 
+        G, F_explicit, w_NF
+    )
+    
+    return nothing
+end
+
+@kernel inbounds=true function accumulate_tendency_kernel!(
+    G, F_explicit, @Const(w)
+)
+    lmk = @index(Global, Linear)
+    G[lmk] = w * F_explicit[lmk] + (1 - w) * G[lmk]
+end
+
+"""$(TYPEDSIGNATURES)
+Update state: x = x + Δt*G (Hotta et al. 2016, Eq. 21)"""
+function update_state!(
+    x::LowerTriangularArray,
+    G::LowerTriangularArray,
+    dt::Real,
+    L::NCycleLorenz{NF},
+) where NF
+    @boundscheck size(x) == size(G) || throw(BoundsError())
+    
+    dt_NF = convert(NF, dt)
+    
+    launch!(
+        architecture(x), 
+        SpectralWorkOrder, 
+        size(x), 
+        update_state_kernel!, 
+        x, G, dt_NF
+    )
+    
+    return nothing
+end
+
+@kernel inbounds=true function update_state_kernel!(
+    x, G, @Const(dt)
+)
+    lmk = @index(Global, Linear)
+    x[lmk] = x[lmk] + dt * G[lmk]
+end
+
+"""$(TYPEDSIGNATURES)
+Lorenz N-cycle time stepping for all prognostic variables following Hotta et al. (2016).
+
+Algorithm per substep:
+1. Accumulate weighted tendency: G = w*F_E + (1-w)*G (Eq. 19)
+2. Apply implicit correction: dx = (I - α*Δt*L_I)^(-1) * (G + L_I*x) (Eq. 20)
+3. Update state: x = x + Δt*dx (Eq. 21)
+"""
+function lorenz_ncycle_timestep!(
+    progn::PrognosticVariables,
+    diagn::DiagnosticVariables,
+    tend::Tendencies,
+    dt::Real,
+    model::AbstractModel,
+)
+    L = model.time_stepping::NCycleLorenz
+    w_current = weight_coefficient(L)
+    
+    # Step 1: Accumulate weighted explicit tendencies (Eq. 19)
+    for (varname, tendname) in zip(prognostic_variables(model), tendency_names(model))
+        var = getfield(progn, varname)
+        
+        # Get G based on dimensionality
+        if ndims(var) == 2
+            # 2D array (e.g., pres in shallow water): direct access
+            G = var[:, 2]
```

Comment:
this is allocating, use `lta_view` for non-allocating and I think we have `get_step` for that!